gcp_variant_transforms/beam_io/vcf_parser.py

```diff
@@ -0,0 +1,455 @@
+        implies the phase of the bases and is consistent with any other variant
+        calls in the same reference sequence which have the same phaseset value.
+        If the genotype data was phased but no phase set was specified, this
+        field will be set to `*`.
+      info: A map of additional variant call information. The key is specified
+        in the VCF record and the type of the value is specified by the VCF
+        header FORMAT.
+    """
+    self.name = name
+    self.genotype = genotype or []
+    self.phaseset = phaseset
+    self.info = info or {}
+
+  def __eq__(self, other):
+    return ((self.name, self.genotype, self.phaseset, self.info) ==
+            (other.name, other.genotype, other.phaseset, other.info))
+
+  def __lt__(self, other):
+    if self.name != other.name:
+      return self.name < other.name
+    elif self.genotype != other.genotype:
+      return self.genotype < other.genotype
+    elif self.phaseset != other.phaseset:
+      return self.phaseset < other.phaseset
+    else:
+      return self.info < other.info
+
+  def __le__(self, other):
+    return self < other or self == other
+
+  def __gt__(self, other):
+    return other < self
+
+  def __ge__(self, other):
+    return other <= self
+
+  def __ne__(self, other):
+    return not self == other
+
+  def __repr__(self):
+    return ', '.join(
+        [str(s) for s in [self.name, self.genotype, self.phaseset, self.info]])
+
+
+class VcfParser(object):
+  """Base abstract class for defining a VCF file parser.
+  Derived classes must implement two methods:
+    _init_with_header: must initialize parser with given header lines.
+    _get_variant: given a line of VCF file, returns a Variant object.
+  Objects of the derived classed will be an iterator of records:
+  ```
+  record_iterator = DerivedVcfParser(...)
+  for record in record_iterator:
+    yield record
+  ```
+  """
+
+  def __init__(self,
+               file_name,  # type: str
+               range_tracker,  # type: range_trackers.OffsetRangeTracker
+               file_pattern,  # type: str
+               compression_type,  # type: str
+               allow_malformed_records,  # type: bool
+               representative_header_lines=None,  # type:  List[str]
+               **kwargs  # type: **str
+              ):
+    # type: (...) -> None
+    # If `representative_header_lines` is given, header lines in `file_name`
+    # are ignored; refer to _process_header_lines() logic.
+    self._representative_header_lines = representative_header_lines
+    self._file_name = file_name
+    self._allow_malformed_records = allow_malformed_records
+
+    text_source = textio._TextSource(
+        file_pattern,
+        0,  # min_bundle_size
+        compression_type,
+        True,  # strip_trailing_newlines
+        coders.StrUtf8Coder(),  # coder
+        validate=False,
+        header_processor_fns=(
+            lambda x: not x or not x.strip() or x.startswith('#'),
```

Comment:
Shouldn't this be `x.strip() or x.startswith('#')`? Otherwise, this would treat the entire file as a header. The unit tests would still pass given that PyVCF does't care about header vs. data lines, but integration tests would fail (likely run out of memory or take way too long for large samples).
Also, you dont need to worry about x being None here.

Reply:
I actually made the same mistake first and later realized it has to be `no x.strip()` otherwise all not empty lines would be considered as header since they all satisfy the `x.strip()` condition.

Reply:
ah, good call! you're right! ...and i'm glad to hear that our unit tests did catch the issue too :)